File: doc/rapport/code/eventsource.java
@Singleton
public class EventSourceController extends Controller implements Observer {

    private static Source<String, ?> source;

    @Override
    public synchronized void update(Observable o, Object arg) {

        if (source == null) {
            source = Source.tick(Duration.ZERO, Duration.ofSeconds(5), "tick");
        }
        else {
            List<String> list = new ArrayList<>();
            list.add((String) arg);
            Source<String, ?> s = Source.from(list);
            source.merge(s);
        }
    }

    public Result events() {

        final Source<EventSource.Event, ?> eventSource;

        return ok().chunked(source
            .map(EventSource.Event::event)
            .via(EventSource.flow()))
            .as(Http.MimeTypes.EVENT_STREAM);

    }
}

File: doc/rapport/code/eventsource.scala
@Singleton
class EventSourceController @Inject() (system: ActorSystem,
                                        cc: ControllerComponents)
                                       (implicit executionContext: ExecutionContext)
extends AbstractController(cc) {

  private[this] val manager = system.actorOf(EventActorManager.props)
  implicit def pair[E]: EventNameExtractor[E] = EventNameExtractor[E](_ => Some("test1"))

  def send(event: String) =  {
    print("Send event to screens : " + event)
    manager ! SendMessage(event)
    Ok
  }

  def index = Action {
    Ok(eventsource.render())
  }

  def events = Action {
    val source  =
      Source
        .actorRef[String](32, OverflowStrategy.dropHead)
        .watchTermination() { case (actorRef, terminate) =>
          manager ! Register(actorRef)
          terminate.onComplete(_ => manager ! UnRegister(actorRef))
          actorRef
        }

    val eventSource = Source.fromGraph(source.map(EventSource.Event.event))
    Ok.chunked(eventSource via EventSource.flow).as(ContentTypes.EVENT_STREAM)
  }
}



